### PR TITLE
Departments nest

### PR DESCRIPTION
**Pull request by mabdullahkhalil**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,8 +1,7 @@
 source 'https://rubygems.org'
 
-
 gem 'autoprefixer-rails', '~> 9.6', '>= 9.6.1'
-gem 'cancancan', '~>3.0.1'
+gem 'cancancan'
```
Comment (zainsaleem4338) [inline Gemfile]: use cancancan 3.0.1 instead

File: app/controllers/departments_controller.rb [hunk 2]
```diff
@@ -1,6 +1,16 @@
 class DepartmentsController < ApplicationController
   load_and_authorize_resource
 
+  def index
+    if !params[:checked].nil? && current_employee.role != 'Admin'
```
Comment (zainsaleem4338) [inline app/controllers/departments_controller.rb]: checked is not so useful name, rename to something better
use 'Admin' as constant (I think it is already defined)

File: app/controllers/departments_controller.rb [hunk 3]
```diff
@@ -1,6 +1,16 @@
...
+      @departments = Department.employee_departments(current_employee)
```
Comment (zainsaleem4338) [inline app/controllers/departments_controller.rb]: load through company

Comment (zainsaleem4338) [reply app/controllers/departments_controller.rb]: infact load through cancancan

File: app/controllers/projects_controller.rb [hunk 4]
```diff
@@ -1,41 +1,47 @@
 class ProjectsController < ApplicationController
-  before_action :set_department, only: [:edit, :index, :new, :create]
-  load_and_authorize_resource through_association: :company
- 
+  load_and_authorize_resource :department
+  load_and_authorize_resource through: :department
+
+  def index
+    if !params[:checked].nil? && current_employee.role != 'Admin'
```
Comment (zainsaleem4338) [inline app/controllers/projects_controller.rb]: same admin constant and  param name change 
and ! nil? and be replace with present

File: app/controllers/projects_controller.rb [hunk 5]
```diff
@@ -1,41 +1,47 @@
...
+      @projects = Project.employee_projects(current_employee)
```
Comment (zainsaleem4338) [inline app/controllers/projects_controller.rb]: load through company.

File: app/models/department.rb [hunk 6]
```diff
@@ -4,4 +4,6 @@ class Department < ActiveRecord::Base
   has_many :teams
   has_many :employees
   validates :name, presence: true
+  scope :get_departments, ->(user){ joins(projects: :department).where(projects: {:id => Project.get_projects(user) }).distinct}
+  scope :employee_departments, ->(user){ joins(projects: :department).where(projects: {:id => Project.employee_projects(user) }).distinct}
```
Comment (zainsaleem4338) [inline app/models/department.rb]: id: instead of :id =>

File: app/models/department.rb [hunk 7]
```diff
@@ -4,4 +4,6 @@ class Department < ActiveRecord::Base
   has_many :teams
   has_many :employees
   validates :name, presence: true
+  scope :get_departments, ->(user){ joins(projects: :department).where(projects: {:id => Project.get_projects(user) }).distinct}
```
Comment (zainsaleem4338) [inline app/models/department.rb]: id:  instead of :id =>

File: app/models/task.rb [hunk 8]
```diff
@@ -7,21 +7,22 @@ class Task < ActiveRecord::Base
   belongs_to :company
   belongs_to :project
   belongs_to :assignable, polymorphic: true
-  validates :complexity
```
Comment (zainsaleem4338) [inline app/models/task.rb]: you removed it! is it being fixed by tehreem?

File: app/views/departments/index.html.erb [hunk 9]
```diff
@@ -1,31 +1,14 @@
...
-        <td><%= link_to 'Show', department, class: "btn btn-regular btn-secondary" %>
-        <%= link_to 'Edit', edit_department_path(department), class: "btn btn-outline-secondary" %>
-        <%= link_to 'Destroy', department, method: :delete, data: { confirm: 'Are you sure?' }, class: "btn btn-outline-danger" %>
-        </td>
-      </tr>
-    <% end %>
-  </tbody>
-</table>
+<div id="departments_grid" >
+	<%= render partial: 'departmentgrid', locals: {departments: @departments}  %>
```
Comment (zainsaleem4338) [inline app/views/departments/index.html.erb]: indentation

File: app/views/projects/index.html.erb [hunk 10]
```diff
@@ -1,26 +1,15 @@
...
-  </tbody>
+<h1>Listing Projects</h1>
+<% if can? :create, Project%>
+<%= link_to 'New Project', new_department_project_path, class: "btn btn-primary" %>
+<% else %>
+    <%= link_to "All Projects", department_projects_path, remote: true, class: "btn btn-secondary active"%>
+    <%= link_to "My Projects", department_projects_path(checked: true), remote: true, class: "btn btn-secondary active" %>
+<% end %>
+<%= link_to 'Back', departments_url, class: "btn btn-primary" %>
+<% if !@projects.nil?%>
```
Comment (zainsaleem4338) [inline app/views/projects/index.html.erb]: projects.present or projects.any if array

File: app/models/task.rb (hunk 8 above)
Comment (mabdullahkhalil) [reply app/models/task.rb]: asked her. she said to remove it until her branch merges at the master

**Comment by mabdullahkhalil** [comment]
Cycle 1 review done

**Pull request merged**